Core: Import STEP: Cancel button does not cancel

supermixed opened the pull request:
Fix for #16282 "Core: Import STEP: Cancel button does not cancel".

This is achieved by throwing a python exception when the STEP options dialog is cancelled (`dlg.exec()` returns 0).

PaddleStroke commented:
I cannot say if throwing a python exception is the correct way of doing it as I've never done this before. @chennes can you please confirm? Thanks

mdrost commented:
I personally am very against using exceptions for control flow. In Visual Studio debugging output already is looking like that:
![obraz](https://github.com/user-attachments/assets/8243cbb8-ba4f-4061-968d-f4f0def03d5e)

chennes commented:
@mdrost is right in that exceptions should be *exceptional* -- that is, not commonplace occurrences. The problem is that the way we currently do imports, this is the simplest way of achieving the cancellation. Anything else will require more significant refactoring, in part because the calling code doesn't really expect you to be popping up a dialog at all, and doesn't have any conception of the idea that the import operation could be cancelled. I'm going to defer to @wwmayer as to whether this PR is an acceptable path forward here, or if we really need to do the refactor instead.

wwmayer commented:
First of all, the bug report is wrong as it claims that pressing OK and Cancel does the same thing. The difference is that when pressing the OK button the settings are saved to the user config.

So far the Cancel button is not supposed to cancel the whole import operation but not to save the dialog settings. If this behaviour should be changed then not doing it the way this PR tries to do.

It's especially a bad thing to use the basic Exception class because for the calling instance it's impossible to distinguish wether a real error has occurred or whether it was caused by user interactions. Additionally, no exception text is passed.

So, pressing the cancel button leads to the weird behaviour that an error message of this form is written:
```
pyException: Traceback (most recent call last):
  File "<string>", line 2, in <module>
<class 'SystemError'>: <built-in method importOptions of tuple object at 0x7f36067f91c0> returned NULL without setting an exception
```
Since cancelling the dialog is a valid operation no error message should be printed.
 
IMO, using exceptions to handle the control flow is not per se a bad thing. If it keeps the business code clean compared to using return values where in the calling instance code must be written that doesn't belong to the business logic then using exceptions can be a valid alternative. However, when doing so then use a specialized sub-class of Exception.

A second regression of this PR is that when the user has activated the option "Don't show this dialog again" then for any subsequent call of importOptions() the method returns an empty dict instead of a dict with the default values.

supermixed commented:
Ok, I think that I've taken everyone's feedback onboard.

As a FreeCAD user, I think it makes sense for the cancel button in this dialog to cancel importing the file.

![image](https://github.com/user-attachments/assets/f38f7da8-20de-4f85-8a03-5933abd9c0b6)

While I get that this goes against the existing behaviour, I think that it makes more sense and reduces user confusion.

I'm sticking with the exception approach here. Why? The primary reason is compatibility.
https://github.com/FreeCAD/FreeCAD/blob/b4923d1c584a4dce794d91a4b090a5293569223a/src/Gui/Application.cpp#L642-L651

I'm assuming that the existing code has been written this way so that it can be extended by other modules e.g. for importing files where support is not built into FreeCAD. This means we probably shouldn't break the existing interface, which is "call the `importOptions` method if it exists, and pass the result of that to the `options` arg of `.insert()` or `.open()`".

As this gives no way to return anything that would indicate the user has cancelled the object insertion (or even that it was done by `.importOptions()` for another reason) we can only really achieve this functionality without breaking the interface by using exceptions. I considered adding an optional second value to be returned from `.importOptions()` but I don't think we can do this without breaking the interface - we don't know if the existing `.importOptions()` of some custom imported already returns multiple values, as we just slam it into `.insert()` etc. Using an exception provides for both forwards and backwards compatibility when using mismatched versions of an imported and FreeCAD.

However, what exception class should we use? I think we should use an exception class that is built into Python, as it prevents adding possibly another interface which could break. I also wasn't sure how to expose a custom exception type to calling code if that is the route we wanted to go down. I used a `RuntimeError`, and we check the message contents to distinguish between the user cancelling the import, and any other kind of runtime error. I wasn't satisfied with this, but couldn't come up with anything better. This should at least fix the issue (or at least what I think is an issue) without polluting the python console (or any console) with exceptions.

I think I have fixed the issue identified by @wwmayer where the default settings were not used (incorrectly) when the user had opted not to show the STEP settings dialog.

I also refactored the python code which we generate dynamically in order to reduce duplication and hopefully make it easier to read. I think that using string interpolation (via `QString::arg`) makes the code a lot more readable than using `std::stringstream` - but I'm not sure if this is the right approach as I am not super-familiar with FreeCAD's codebase.

wwmayer commented:
> As this gives no way to return anything that would indicate the user has cancelled the object insertion 

Theoretically the returned options object can be checked whether it's False and then abort the import. But in this case I also like it more to use an exception because it's more explicit and checking for False now is a reinterpretation of the returned options object.

> However, what exception class should we use?

We are free to define our own exception types in Python. Just have a look at PyObjectBase.h. E.g. it can be called `PyExc_FC_AbortIOException`. It's then implemented in `Application::setupPythonException` where on Python side the class name could be `AbortIOException`.

> I think we should use an exception class that is built into Python.

When I have a look at https://docs.python.org/3/library/exceptions.html I couldn't find a type that could be used for our use purposes without causing ambiguities.

> I used a RuntimeError, and we check the message contents to distinguish between the user cancelling the import, and any other kind of runtime error.

This is exactly what should be avoided because the code can break easily.

> I also refactored the python code which we generate dynamically in order to reduce duplication and hopefully make it easier to read. I think that using string interpolation (via QString::arg) makes the code a lot more readable than using std::stringstream - but I'm not sure if this is the right approach as I am not super-familiar with FreeCAD's codebase

In the past I more and more moved embedded Python code to real modules because writing and testing the code will be much simpler. E.g. in src/Ext/freecad we can add a new module module_io.py or so, move there the relevant Python code and then import it in the C++ code using the PyCXX API.

supermixed commented:
@wwmayer I have added the custom exception type and pulled out the python code to a new file like you suggested.

wwmayer commented:
Thanks a lot! It looks very close to what I had in mind.

supermixed commented:
No worries - I learned a lot!

bgbsww commented on src/Gui/Application.cpp:
```diff
@@ -636,19 +637,21 @@ void Application::open(const char* FileName, const char* Module)
-                    << "    " << Module << ".open(u\"" << unicodepath << "\")\n";
-
-                std::string code = str.str();
-                Gui::Command::runCommand(Gui::Command::App, code.c_str());
+                // Load using provided python module
+                {
+                    Base::PyGILStateLocker locker;
+                    Py::Module moduleIo(PyImport_ImportModule("freecad.module_io"));
+                    const auto dictS = moduleIo.getDict().keys().as_string();
+                    if (!moduleIo.isNull() && moduleIo.hasAttr("OpenInsertObject"))
+                    {
+                        const Py::TupleN args(
+                            Py::Module(PyImport_ImportModule(Module)),
+                            Py::String(unicodepath),
+                            Py::String("open")
+                        );
+                        moduleIo.callMemberFunction("OpenInsertObject", args);
+                    }
+                }
```
Pulling out the doCommand and runCommand here means that the actions taken will no longer be logged to the python console, and can no longer be recorded in a macro.  Is that intentional?

supermixed replied:
You're right, and that wasn't intentional. I wrote it that way because @wwmayer above said to use the PyCXX API - I didn't see a way to square that with using the `Gui::Command` class. Is there a way to have structured C++ code (e.g. using PyCXX rather than mashing strings together) and also have the commands visible in the console/macroable? Should we revert this part of the change? I had a cleaned up version of the string construction that used `QString::arg` in a prior commit - that wouldn't have this regression.

https://github.com/FreeCAD/FreeCAD/blob/de122b6dfa1555b214bb9eaece6153d832109d02/src/Gui/Application.cpp#L601-L627

(yes I messed up the exception name here)

Or do we not care? Or is there another way?

bgbsww replied:
Well, I added doConsoleSkip recently to support cases where you want to put something in the console/macro that is a python equivalent without executing it.  But you need to have the python equivalent available, so in this case you would likely need to move back towards the python.  I can't answer the do we care question - I think so, but maybe we're moving in a different direction?

bgbsww commented on src/Gui/Application.cpp:
```diff
@@ -713,31 +716,23 @@ void Application::importFrom(const char* FileName, const char* DocName, const ch
 
-                std::string code = str.str();
-                Gui::Command::runCommand(Gui::Command::App, code.c_str());
-
```
Same problem.

bgbsww reviewed (commented):
Just noticed these points, if you're in looking at this, maybe these too?

wwmayer commented:
https://github.com/FreeCAD/FreeCAD/pull/16872

maxwxyz commented:
I cannot import STEP files in the current version, could it be that this PR broke it? https://github.com/FreeCAD/FreeCAD/issues/16886
